### ISSUE TITLE
Thyra::DetachedVectorView warnings in gcc 6.2

### ISSUE THREAD

**Issue by rppawlo** [trilinos/Trilinos#611]
@trilinos/thyra 

The Thyra::DetachedVectorView is generating warnings with gcc 6.2 compiler due to exceptions being called in the destructor:

[ 98%] Building CXX object tempus/test/TestModels/CMakeFiles/tempus_test_models.dir/CDR_Model.cpp.o
In file included from /Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_Assert.hpp:46:0,
                 from /Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_Array.hpp:50,
                 from /Users/rppawlo/Trilinos/packages/rtop/src/interfaces/RTOpPack_Types.hpp:49,
                 from /Users/rppawlo/Trilinos/packages/thyra/core/src/interfaces/operator_vector/fundamental/Thyra_OperatorVectorTypes.hpp:46,
                 from /Users/rppawlo/Trilinos/packages/thyra/core/src/interfaces/operator_vector/fundamental/Thyra_VectorBase.hpp:46,
                 from /Users/rppawlo/Trilinos/packages/thyra/core/src/support/nonlinear/model_evaluator/client_support/Thyra_ModelEvaluatorDefaultBase.hpp:45,
                 from /Users/rppawlo/Trilinos/packages/thyra/core/src/support/nonlinear/model_evaluator/client_support/Thyra_StateFuncModelEvaluatorBase.hpp:45,
                 from /Users/rppawlo/Trilinos/tempus/test/TestModels/CDR_Model_decl.hpp:4,
                 from /Users/rppawlo/BUILD/tempus/test/TestModels/CDR_Model.hpp:1,
                 from /Users/rppawlo/Trilinos/tempus/test/TestModels/CDR_Model.cpp:4:
/Users/rppawlo/Trilinos/packages/thyra/core/src/support/operator_vector/client_support/Thyra_DetachedVectorView.hpp: In instantiation of 'Thyra::DetachedVectorView<Scalar>::~DetachedVectorView() [with Scalar = double]':
/Users/rppawlo/Trilinos/tempus/test/TestModels/CDR_Model_impl.hpp:184:43:   required from 'void Tempus_Test::CDR_Model<Scalar>::set_x0(const Teuchos::ArrayView<const T>&) [with Scalar = double]'
/Users/rppawlo/Trilinos/tempus/test/TestModels/CDR_Model.cpp:8:3:   required from here
/Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr); \
                            ^
/Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:325:3: note: in expansion of macro TEUCHOS_TEST_FOR_EXCEPTION'
   TEUCHOS_TEST_FOR_EXCEPTION(throw_exception_test, std::logic_error, msg)
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/rppawlo/Trilinos/packages/thyra/core/src/support/operator_vector/client_support/Thyra_DetachedVectorView.hpp:345:9: note: in expansion of macro 'TEUCHOS_TEST_FOR_EXCEPT_MSG'
         TEUCHOS_TEST_FOR_EXCEPT_MSG(true, "I don't think non-unit stride has ever been tested!");
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: note: in C++11 destructors default to noexcept
     throw Exception(omsgstr); \
                            ^
/Users/rppawlo/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:325:3: note: in expansion of macro TEUCHOS_TEST_FOR_EXCEPTION'
   TEUCHOS_TEST_FOR_EXCEPTION(throw_exception_test, std::logic_error, msg)
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/rppawlo/Trilinos/packages/thyra/core/src/support/operator_vector/client_support/Thyra_DetachedVectorView.hpp:345:9: note: in expansion of macro 'TEUCHOS_TEST_FOR_EXCEPT_MSG'
         TEUCHOS_TEST_FOR_EXCEPT_MSG(true, "I don't think non-unit stride has ever been tested!");
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by mhoemmen** [comment]
This isn't _too_ bad.  http://en.cppreference.com/w/cpp/language/noexcept_spec says:  "If a search for a matching exception handler leaves a function marked noexcept or noexcept(true), std::terminate is called immediately."  Thus, it's not undefined behavior for a function marked noexcept to throw.

**Comment by bartlettroscoe** [comment]
It is bad form in general to allow exceptions to leave destructors.  The thread-safe implementation of the Teuchos Memory Management classes will no longer provide the "strong guarantee" when exceptions are thrown (see #229).  Therefore, we should work to stop all exception from from leaving destructores.

It think the way we need to handle these is to create a new Teuchos macro called something like `TEUCHOS_CATCH_ALL_PRINT_AND_ABORT()` (which would actually perhaps call `TEUCHOS_STANDARD_CATCH_STATEMENTS()` or at least use the same guts) and then use it like:

```
SomeClass::~SomeClass()
{
  try {
    ...
  }
  TEUCHOS_CATCH_ALL_PRINT_AND_ABORT(<some-fancyostream-ptr>);
}
```

If an exception is thrown (any exception), then it will be caught and reported to the passed in `*<some-fancyostream-ptr>` object (or `*Teuchos::VerboseObjectBase::getDefaultOStream()` by default if `Teuchos::null` is passed in).

That way we would still get nice trackbacks if stack tracing was enabled and in parallel but we would not get these compiler warning.

What do people think about that solution?

Can we get SEMS to add GCC 6.2 to the list of supported compilers and then set up a Trilinos nightly build to support this?

**Comment by mhoemmen** [comment]
> What do people think about that solution?

I think it's reasonable in general.  Take a look at the specific case, though:
```
TEUCHOS_TEST_FOR_EXCEPT_MSG(true, "I don't think non-unit stride has ever been tested!");
```
This smells like something that the constructor should check.  If the constructor has already checked this case, then the destructor just plain doesn't need to check.

**Comment by bartlettroscoe** [comment]
> This smells like something that the constructor should check. If the constructor has already checked this case, then the destructor just plain doesn't need to check.

I would have to look at the full context for this but my guess is that this is being (perhaps) overly careful.  But given undefined behavior and crazy memory usage bugs, having these types of checks is not a bad idea in a debug-version of the code.

**Comment by ibaned** [comment]
I was opening a similar issue on Teuchos, and then noticed this one. I think we need to come up with a Trilinos-wide decision about how we're going to handle errors in destructors, and #1303 could serve to track that. @bartlettroscoe 's proposed `TEUCHOS_CATCH_ALL_PRINT_AND_ABORT` is fine with me.

**Comment by bartlettroscoe** [comment]
Just curious, but what is driving the usage of GCC 6.2 at this time?  Are there any internal Sandia Trilinos customers that are using that compiler?

**Comment by ibaned** [comment]
@bartlettroscoe on my end, there are two drivers for GCC 6.x series usage:

1. The CEE lan SEMS environment has a GCC 6.1 module, that is what the Alexa shock hydro application selected for its nightly testing on CEE machines.
2. I have an Ubuntu workstation used for open-source work, and the latest version of Ubuntu uses GCC 6.3 as its default GCC version.

Looking beyond that, the Albany project has even started to use GCC 7.1 for some of its nightly builds. For better or worse (I think better), Albany and Alexa will likely be reporting issues with very new GCC versions.
I have no idea what @rppawlo 's drivers are though.

**Comment by rppawlo** [comment]
I use gcc 6.X for daily development work. Still use the sems modules to push with checkin-test script though. This ticket is not high priority, but wanted to point out the issue since it would take some work to fix.

**Comment by bradbell** [comment]
I am getting this warning as a Sacado user. For example,
Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr);
and would like to see this fixed. 

Perhaps cmake could add the proper flags to the compiler so it did not warn about cases the developers know are not problems. For example, in the case of gcc can the -Wno-terminate flag could be used for this purpose ?

**Comment by ibaned** [comment]
@bradbell I agree, and it may be best to comment in #1303, which encompasses this issue and others.

The problem I have with the `-Wno-terminate` flag is that it would never get removed, and the underlying problem would never get fixed, because no one would have the incentive to do it.

**Comment by bradbell** [comment]
Perhaps this should be changd to a discussion, instead of closing it, because it has not been resolved ?

**Comment by bartlettroscoe** [comment]
@bradbell, I believe this was fixed a long time ago by the merge of PR #2089.  There are builds of Trilinos with GCC 8.3.0 that are warning free.
